File: app/src/main/java/org/wikipedia/dataclient/wikidata/Entities.java
package org.wikipedia.dataclient.wikidata;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.JsonElement;
import com.google.gson.annotations.SerializedName;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.dataclient.mwapi.MwResponse;
import org.wikipedia.json.PostProcessingTypeAdapter;

import java.util.Collections;
import java.util.List;
import java.util.Map;

@SuppressWarnings("unused")
public class Entities extends MwResponse implements PostProcessingTypeAdapter.PostProcessable {
    @Nullable private Map<String, Entity> entities;
    private int success;

    @NonNull public Map<String, Entity> entities() {
        return entities != null ? entities : Collections.emptyMap();
    }

    @Nullable public Entity getFirst() {
        if (entities == null) {
            return null;
        }
        return entities.values().iterator().next();
    }

    @Override
    public void postProcess() {
        if (getFirst() != null && getFirst().isMissing()) {
            throw new RuntimeException("The requested entity was not found.");
        }
    }

    public static class Entity {
        @Nullable private String type;
        @Nullable private String id;
        @Nullable private Map<String, Label> labels;
        @Nullable private Map<String, Label> descriptions;
        @Nullable private Map<String, SiteLink> sitelinks;
        @Nullable private Map<String, List<Claim>> claims;
        @Nullable private String missing;
        private long lastrevid;

        @NonNull public String id() {
            return StringUtils.defaultString(id);
        }

        @NonNull public Map<String, Label> labels() {
            return labels != null ? labels : Collections.emptyMap();
        }

        @NonNull public Map<String, Label> descriptions() {
            return descriptions != null ? descriptions : Collections.emptyMap();
        }

        @NonNull public Map<String, SiteLink> sitelinks() {
            return sitelinks != null ? sitelinks : Collections.emptyMap();
        }

        @NonNull public Map<String, List<Claim>> claims() {
            return claims != null ? claims : Collections.emptyMap();
        }

        @NonNull public String getLabelForLang(@NonNull String lang) {
            return (labels != null && labels.containsKey(lang)) ? labels.get(lang).value() : "";
        }

        public boolean isMissing() {
            return "-1".equals(id) && missing != null;
        }

        public long getLastRevId() {
            return lastrevid;
        }
    }

    public static class Label {
        @Nullable private String language;
        @Nullable private String value;

        @NonNull public String language() {
            return StringUtils.defaultString(language);
        }

        @NonNull public String value() {
            return StringUtils.defaultString(value);
        }
    }

    public static class SiteLink {
        @Nullable private String site;
        @Nullable private String title;

        @NonNull public String getSite() {
            return StringUtils.defaultString(site);
        }

        @NonNull public String getTitle() {
            return StringUtils.defaultString(title);
        }
    }

    public static class Claim {
        @Nullable private String type;
        @Nullable private String id;
        @Nullable private String rank;
        @Nullable private Mainsnak mainsnak;

        @NonNull public String getRank() {
            return StringUtils.defaultString(rank);
        }

        @NonNull public String getType() {
            return StringUtils.defaultString(type);
        }

        @Nullable public Mainsnak getMainsnak() {
            return mainsnak;
        }
    }

    public static class Mainsnak {
        @Nullable private String snaktype;
        @Nullable private String datatype;
        @Nullable private String property;
        @Nullable private DataValue datavalue;

        @NonNull public String getSnakType() {
            return StringUtils.defaultString(snaktype);
        }

        @NonNull public String getDataType() {
            return StringUtils.defaultString(datatype);
        }

        @NonNull public String getProperty() {
            return StringUtils.defaultString(property);
        }

        @Nullable public DataValue getDataValue() {
            return datavalue;
        }
    }

    public static class DataValue {
        @Nullable private String type;
        @Nullable private JsonElement value;

        @NonNull public String getType() {
            return StringUtils.defaultString(type);
        }

        @Nullable public JsonElement getValue() {
            return value;
        }
    }

    public static class EntityIdValue {
        @SerializedName("entity-type") @Nullable private String entityType;
        @SerializedName("numeric-id") private int numericId;

        @NonNull public String getEntityType() {
            return StringUtils.defaultString(entityType);
        }

        public int getNumericId() {
            return numericId;
        }
    }

    public static class QuantityValue {
        @Nullable private String amount;
        @Nullable private String unit;
        @Nullable private String lowerBound;
        @Nullable private String upperBound;

        @NonNull public String getAmount() {
            return StringUtils.defaultString(amount);
        }
    }

    public static class TimeValue {
        @Nullable private String time;
        private int timezone;
        private int before;
        private int after;
        private int precision;
        @Nullable private String calendarModel;

        @NonNull public String getTime() {
            return StringUtils.defaultString(time);
        }

        public int getPrecision() {
            return precision;
        }
    }

    public static class LocationValue {
        private float latitude;
        private float longitude;
        private float altitude;
        private float precision;

        public float getLatitude() {
            return latitude;
        }

        public float getLongitude() {
            return longitude;
        }
    }

    public static class MonolingualTextValue {
        @Nullable private String language;
        @Nullable private String text;

        @NonNull public String getText() {
            return StringUtils.defaultString(text);
        }
    }
}

File: app/src/main/java/org/wikipedia/dataclient/wikidata/Entities.kt
package org.wikipedia.dataclient.wikidata

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.decodeFromJsonElement
import org.wikipedia.dataclient.mwapi.MwResponse
import org.wikipedia.json.JsonUtil

@Serializable
class Entities : MwResponse() {

    var entities: Map<String, Entity> = emptyMap()
        private set
    val first: Entity?
        get() = entities.values.firstOrNull()

    init {
        entities = entities.filter { it.key != "-1" && it.value.missing == null }
    }

    @Serializable
    class Entity {
        val id: String = ""
        private val labels: JsonElement? = null
        private val descriptions: JsonElement? = null
        private val sitelinks: JsonElement? = null
        private val statements: JsonElement? = null
        val missing: JsonElement? = null
        @SerialName("lastrevid") val lastRevId: Long = 0

        fun getStatements(): Map<String, List<Claims.Claim>> {
            return if (statements != null && statements !is JsonArray) {
                JsonUtil.json.decodeFromJsonElement(statements)
            } else {
                emptyMap()
            }
        }

        val claims: Map<String, List<Claim>> = emptyMap()

        fun getLabelForLang(lang: String): String {
            return getLabels()[lang]?.value.orEmpty()
        }

        fun getLabels(): Map<String, Label> {
            return if (labels != null && labels !is JsonArray) {
                JsonUtil.json.decodeFromJsonElement(labels)
            } else {
                emptyMap()
            }
        }

        fun getDescriptions(): Map<String, Label> {
            return if (descriptions != null && descriptions !is JsonArray) {
                JsonUtil.json.decodeFromJsonElement(descriptions)
            } else {
                emptyMap()
            }
        }

        fun getSiteLinks(): Map<String, SiteLink> {
            return if (sitelinks != null && sitelinks !is JsonArray) {
                JsonUtil.json.decodeFromJsonElement(sitelinks)
            } else {
                emptyMap()
            }
        }
    }

    @Serializable
    class Label {
        val language: String = ""
        val value: String = ""
    }

    @Serializable
    class SiteLink {
        val site: String = ""
        val title: String = ""
    }

    @Serializable
    class Claim {
        val type: String? = null
            get() = field.orEmpty()

        private val id: String = ""
        val rank: String = ""
        val mainsnak: Mainsnak? = null
    }

    @Serializable
    class Mainsnak {
        val snaktype: String = ""
        val datatype: String = ""
        val property: String = ""
        val datavalue: DataValue? = null
    }

    @Serializable
    class DataValue {
        val type: String = ""
        val value: JsonElement? = null
    }

    @Serializable
    class EntityIdValue {
        @SerialName("entity-type")
        val entityType: String = ""
        @SerialName("numeric-id")
        val numericId = 0
    }

    @Serializable
    class QuantityValue {
        val amount: String = ""
        val unit: String = ""
        val lowerBound: String = ""
        val upperBound: String = ""
    }

    @Serializable
    class TimeValue {
        val time: String = ""
        private val timezone = 0
        private val before = 0
        private val after = 0
        val precision = 0
        val calendarModel: String = ""
    }

    @Serializable
    class LocationValue {
        val latitude = 0f
        val longitude = 0f
        private val altitude = 0f
        private val precision = 0f
    }

    @Serializable
    class MonolingualTextValue {
        private val language: String? = null
        val text: String = ""
    }
}
